Add sample playback support for Web exports

adamscott opened the pull request:
# ⚠️ Here be dragons ⚠️
When this PR will be merged for 4.3, this feature will be marked as experimental and could change in 4.4.

# Based on these PRs
- https://github.com/godotengine/godot/pull/92568

# tl;dr

This PR adds (back) the concept of samples to the Godot Engine. Currently, the PR enables only the web platform to play samples.

It principally fixes #87329, as that issue would plague any non-threaded web releases with crackling audio.

## Example
| Single-threaded web example using streams<br>(old way) | Single-threaded web example using samples<br>(this PR) |
| --- | --- |
| https://adamscott.github.io/2d-platformer-demo-main-thread/ | https://adamscott.github.io/2d-platformer-demo-main-thread-samples/ |

# Introduction

Godot uses streaming to mix game audio. Each active stream is registered and then the engine mix on-the-fly the needed audio frames together to output audio based on the audio latency parameter. It works very well on modern platforms.

Samples are another way to handle sound instead of mixing streams. Instead of handling mixing sound and music by the game processes, it relies on off-loading it to the host system. While it doesn't permit full access to the mixing apparatus, it's super useful on systems that don't have a lot of processing power.

To use samples, you register a sample, and then tell the system to play it when needed. And to stop it. It's like a music player, you set the file, then you click on play. You don't control how the software do it, but you know it does.

Godot used to have samples back in Godot 1 and 2, especially to support platforms like the PSP, and the web (thanks to the [Web Audio API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Audio_API)).

As newer console platforms let developers handle their own mixing logic and that `SharedArrayBuffer`s were introduced in browsers (permitting WebWorkers (web threads) to share memory with the game), samples support was dropped from Godot. Everything was fine.

Anyway, the implementation was somewhat lacking. You had to specifically want to play samples, you couldn't use common nodes to play both streams and samples.

# The problem
But on the web platform, Spectre and Meltdown happened. And it completely changed where `SharedArrayBuffer`s were able to be used. [Enter "cross-origin isolated" websites](https://web.dev/articles/coop-coep), where it's impossible to contact other websites or display ads, and complicating hosting of simple games, greatly reducing the appeal for our web builds.

Hence the work on #85939 in order to compile Godot to run on the main thread. This enables exporting Godot games on the browser without having to cross-origin isolate your website. Unfortunately, this brought an unexpected issue: software mixing is pretty much incompatible with single-threaded games. Especially running on older/less powerful hardware.

Wanna hear for yourself? Try the [single-threaded platformer demo (without this PR applied)](https://adamscott.github.io/2d-platformer-demo-main-thread/) on your phone or on a computer that doesn't have a great CPU.

# The investigation
My colleague @Faless and I considered every solution imaginable: augmenting latency for the web, traced the processes on the web and on mobile and refactor the `AudioWorklet` processing the audio. But alas. Nothing substantial could have been done.

The only solution we found was to resort to Web Audio samples.

And it's not uncommon for web game engines. _We were the uncommon ones not using web audio samples._ So, a few weeks ago, I began work on this PR. 

# The sole requirement: seamlessness
My main focus was to reuse as possible as many features that already exist. It means that in order to play samples, I wanted the UX to keep as close as possible to existing tools.

Godot strives itself to offer the same experience for every target that it exports to. Imagine making the developer choose between having samples for the web export and streaming nodes for the rest. And having to manually add or remove nodes based on the platform with scripts.

This has such a big impact that it's a clear no go for us. We don't want that poor UX.

# The solution
My solution is a big hack. (But it does works wonderfully.)

The idea is to reuse all the existing stream nodes and systems. And make the stream elements capable of producing samples.

This story begins with the new project setting `audio/general/default_playback_type` (hidden currently in the advanced options). Usually, it should stay with the value "Stream", as normally, that's how Godot works currently. But the magic happens with `audio/general/default_playback_type.web` set as "Sample".

That's because `AudioStreamPlayer`, `AudioStreamPlayer2D` and `AudioStreamPlayer3D` now have a new property called `playback_type`, which is set by default to... "Default". That's where the magic happens! On standard exports, the nodes will be defined as "Stream", but on web exports, "Sample" will be used instead!

The magic operates behind the scenes though. 

## The man behind the curtain
Essentially, when a stream is considered a "sample", it doesn't get mixed at all in the mixing phase. Instead, it relies on callbacks by the `StreamPlayer` nodes.

The `StreamPlayer` nodes, when their `play()` method is called, are calling internally `AudioServer::start_sample_playback()`. All the `AudioServer` does is to call `AudioDriver::start_sample_playback()`. If the driver doesn't implement that function, it just doesn't play any sound. But if it does, the driver can now tell the backend to play that sound.

The same thing happens for stop, pause, etc. You can even update the sample, like when the position of the node changes!

Isn't this fascinating?

## Registering samples
Before playing the samples, it's important to register them first.

If played without previous registration, the player will make sure to register it first. Though, it's recommended to register manually streams. That's because, on single threaded games, memory transfer is synchronous, so it may make your game stutter. You register a stream as a sample by calling this method:

```gdscript
# optional step
const my_stream_resource = preload("res://assets/my_stream.wav")

AudioServer.register_stream_as_sample(my_stream_resource)
```

Under the hood, Godot will call the `mix()` method of the stream playback for the entire duration of the clip. This makes it so that it's possible to play any type of sound media that Godot supports (WAV, mp3, ogg vorbis).

# Is it really seamless, though?

These demos were exported to the web (single-threaded) using samples without ever touching the project nodes, resources, nor files.

| Demo title | Playable link | Source |
| --- | --- | --- |
| 2D Platformer | [Play](https://adamscott.github.io/2d-platformer-demo-main-thread-samples/) | [Link](https://github.com/godotengine/godot-demo-projects/tree/master/2d/platformer) |
| Dodge the Creeps | [Play](https://adamscott.github.io/2d-dodge-the-creeps-demo-main-thread-samples) | [Link](https://github.com/godotengine/godot-demo-projects/tree/master/2d/dodge_the_creeps) |
| 3D Platformer | [Play](https://adamscott.github.io/3d-platformer-demo-main-thread-samples/) | [Link](https://github.com/godotengine/godot-demo-projects/tree/master/3d/platformer) |
| Truck Town | [Play](https://adamscott.github.io/truck-town-main-thread-samples/) | [Link](https://github.com/godotengine/godot-demo-projects/tree/master/3d/truck_town) |
| Hell of Mirrors | [Play](https://adamscott.github.io/hell-of-mirrors-main-thread-samples/) | [Link](https://bliznyuknm.itch.io/hell-of-mirrors) |
| Catburglar | [Play](https://adamscott.github.io/catburglar-main-thread-samples/) | [Link](https://johngabrieluk.itch.io/catburglar) |

# Bugs yet to fix before merge

- [x] ~~Buses don't chain properly~~
- [x] ~~Autoplay doesn't work right now.~~
- [x] ~~Fix issues with sample rate.~~
- [x] ~~AnimationPlayer cannot play samples.~~
- [x] ~~Advanced audio importers fail to show / infinite loop~~ _(the problem came and go without any of my input)_
- [ ] ~~Only forward loop is supported (fix may not make it to the final release~~) _looping is kinda a little broken right now_

# Known limitations
- Effects don't apply (will certainly not be part of the initial release and GDScript based effects cannot be used)

# Technical diagrams

## Registering and playing samples
```mermaid
sequenceDiagram
    participant Script
    participant AudioStreamPlayer2D
    participant AudioStreamPlayerInternal
    participant AudioServer
    participant AudioDriverWeb
    participant JavaScriptAudioLibrary

    Script ->> AudioStreamPlayer2D: set_stream(Ref<AudioStream>)
    AudioStreamPlayer2D ->> AudioStreamPlayerInternal: set_stream(Ref<AudioStream>)

    Script ->> AudioServer: register_stream_as_sample(Ref<AudioStream>)
    activate AudioServer
    AudioServer ->> AudioServer: register_sample(Ref<AudioSample>)
    AudioServer ->> AudioDriverWeb: register_sample(Ref<AudioSample>)
    deactivate AudioServer
    AudioDriverWeb ->> JavaScriptAudioLibrary: godot_audio_sample_register_stream()

    Script ->> AudioStreamPlayer2D: play()
    activate AudioStreamPlayer2D
    AudioStreamPlayer2D ->> AudioStreamPlayerInternal: play_basic()
    activate AudioStreamPlayerInternal
    opt _is_sample() && stream->can_be_sampled() && stream_playback->sample_playback is null
        AudioStreamPlayerInternal ->> AudioServer: is_stream_registered_as_sample(Ref<AudioStream>)
        activate AudioServer
        deactivate AudioServer
        AudioServer ->> AudioStreamPlayerInternal: 
        opt stream is not registered
            AudioStreamPlayerInternal ->> AudioServer: register_stream_as_sample(Ref<AudioStream>)
            activate AudioServer
            AudioServer ->> AudioServer: register_sample(Ref<AudioSample>)
            AudioServer ->> AudioDriverWeb: register_sample(Ref<AudioSample>)
            deactivate AudioServer
            AudioDriverWeb ->> JavaScriptAudioLibrary: godot_audio_sample_register_stream()
        end
        AudioStreamPlayerInternal ->> AudioStreamPlayerInternal: set stream_playback->sample_playback
    end
    AudioStreamPlayerInternal ->> AudioStreamPlayer2D: Ref<AudioStreamPlayback>
    
    deactivate AudioStreamPlayerInternal

    opt stream playback sample exists
        AudioStreamPlayer2D ->> AudioStreamPlayer2D: Update stream_playback->sample_playback for specific AudioStreamPlayer2D stuff
        AudioStreamPlayer2D ->> AudioServer: start_sample_playback(Ref<AudioSamplePlayback>)
        deactivate AudioStreamPlayer2D

        AudioServer ->> AudioDriverWeb: start_sample_playback(Ref<AudioSamplePlayback>)
        AudioDriverWeb ->> JavaScriptAudioLibrary: godot_audio_sample_start()
    end
```

## Samples and streams
```mermaid
classDiagram
    namespace AudioStreams {
        class AudioStream
        class AudioStreamWAV
        class AudioStreamPlayback
        class AudioStreamPlaybackWAV
    }
    namespace AudioSamples {
        class AudioSample
        class AudioSamplePlayback
    }

    %% Inheritance
    AudioStream <|-- AudioStreamWAV
    AudioStreamPlayback <|-- AudioStreamPlaybackWAV

    %% Links
    AudioStream ..|> AudioSample
    AudioStreamWAV ..|> AudioSample

    AudioStreamPlayback ..|> AudioSamplePlayback

    %% Cardinality
    AudioStreamPlaybackWAV "1" --> "many" AudioSamplePlayback

    AudioSample "1" --> "many" AudioStream
    AudioSamplePlayback "1" --> "many" AudioStream

    %% Classes
    class AudioStream {
        +can_be_sampled()*
        +get_sample()
    }

    class AudioStreamWAV {
        +can_be_sampled()
        +get_sample()
    }

    class AudioStreamPlayback {
        +set_is_sample()*
        +get_is_sample()*
        +set_sample_playback()*
        +get_sample_playback()*
    }

    class AudioStreamPlaybackWAV {
        -bool _is_sample
	    -Ref~AudioSamplePlayback~ sample_playback

        +set_is_sample()
        +get_is_sample()
        +set_sample_playback()
        +get_sample_playback()
    }

    class AudioSample {
        +Ref~AudioStream~stream
        +data
        +num_channels
        +sample_rate
        +loop_mode
        +loop_begin
        +loop_end
    }

    class AudioSamplePlayback {
        +Ref~AudioStream~stream
        +float offset
        +float volume_db
        +PositionMode position_mode
        +Vector3 position
        +StringName bus
    }
```

# Fixes
Fixes #87329

adamscott committed:
Add samples playback support

AThousandShips commented on modules/minimp3/audio_stream_mp3.h:
```diff
@@ -131,6 +139,11 @@ class AudioStreamMP3 : public AudioStream {
 
+	virtual bool can_be_sampled() const override {
+		return true;
+	};
```
```suggestion
	}
```

AThousandShips commented on modules/vorbis/audio_stream_ogg_vorbis.h:
```diff
@@ -159,6 +167,11 @@ class AudioStreamOggVorbis : public AudioStream {
 
+	virtual bool can_be_sampled() const override {
+		return true;
+	};
```
```suggestion
	}
```

AThousandShips commented on scene/audio/audio_stream_player_internal.h:
```diff
@@ -51,12 +54,17 @@ class AudioStreamPlayerInternal : public Object {
 
+	_FORCE_INLINE_ bool _is_sample() {
+		return (AudioServer::get_singleton()->get_default_playback_type() == AudioServer::PlaybackType::PLAYBACK_TYPE_SAMPLE && get_playback_type() == AudioServer::PlaybackType::PLAYBACK_TYPE_DEFAULT) || get_playback_type() == AudioServer::PlaybackType::PLAYBACK_TYPE_SAMPLE;
+	};
```
```suggestion
	}
```

AThousandShips commented on servers/audio_server.cpp:
```diff
@@ -1705,6 +1724,82 @@ void AudioServer::get_argument_options(const StringName &p_function, int p_idx,
+
+void AudioServer::unregister_stream_as_sample(const Ref<AudioStream> &p_stream) {
+	ERR_FAIL_COND_MSG(p_stream.is_null(), "Parameter p_stream is null.");
+	ERR_FAIL_COND_MSG(!(p_stream->can_be_sampled()), "Parameter p_stream cannot be sampled.");
```
```suggestion
	ERR_FAIL_COND_MSG(!p_stream->can_be_sampled(), "Parameter p_stream cannot be sampled.");
```

AThousandShips commented on servers/audio_server.cpp:
```diff
@@ -1705,6 +1724,82 @@ void AudioServer::get_argument_options(const StringName &p_function, int p_idx,
+void AudioServer::register_sample(const Ref<AudioSample> &p_sample) {
+	ERR_FAIL_COND_MSG(p_sample.is_null(), "Parameter p_sample is null.");
+	ERR_FAIL_COND_MSG(p_sample->stream.is_null(), "Parameter p_sample->stream is null.");
+	ERR_FAIL_COND_MSG(!(p_sample->stream->can_be_sampled()), "Parameter p_stream cannot be sampled.");
```
```suggestion
	ERR_FAIL_COND_MSG(!p_sample->stream->can_be_sampled(), "Parameter p_stream cannot be sampled.");
```

adamscott replied:
I had errors with the macro if I didn't use the parentheses, hence why they are there.

AThousandShips replied:
Interesting, that's something to look into, this should be valid

AThousandShips commented on servers/audio/audio_stream.h:
```diff
@@ -75,6 +122,14 @@ class AudioStreamPlayback : public RefCounted {
 
 	virtual int mix(AudioFrame *p_buffer, float p_rate_scale, int p_frames);
+
+	virtual void set_is_sample(bool p_is_sample){};
+	virtual bool get_is_sample() const { return false; };
+	virtual Ref<AudioSamplePlayback> get_sample_playback() const;
+	virtual void set_sample_playback(const Ref<AudioSamplePlayback> &p_playback){};
```
```suggestion
	virtual void set_is_sample(bool p_is_sample) {}
	virtual bool get_is_sample() const { return false; }
	virtual Ref<AudioSamplePlayback> get_sample_playback() const;
	virtual void set_sample_playback(const Ref<AudioSamplePlayback> &p_playback) {}
```

AThousandShips commented on servers/audio_server.h:
```diff
@@ -129,6 +131,26 @@ class AudioDriver {
+	// Samples handling.
+	virtual bool is_stream_registered_as_sample(const Ref<AudioStream> &p_stream) const {
+		return false;
+	};
+	virtual void register_sample(const Ref<AudioSample> &p_sample){};
+	virtual void unregister_sample(const Ref<AudioSample> &p_sample){};
+	virtual void start_sample_playback(const Ref<AudioSamplePlayback> &p_playback){};
+	virtual void stop_sample_playback(const Ref<AudioSamplePlayback> &p_playback){};
+	virtual void set_sample_playback_pause(const Ref<AudioSamplePlayback> &p_playback, bool p_paused){};
+	virtual bool is_sample_playback_active(const Ref<AudioSamplePlayback> &p_playback) { return false; };
+	virtual void update_sample_playback(const Ref<AudioSamplePlayback> &p_playback, const StringName &p_bus, float p_pan = 0.0f, float p_volume_db = 0.0f, float p_pitch_scale = 0.0f){};
+
+	virtual void set_sample_bus_count(int p_count){};
+	virtual void remove_sample_bus(int p_bus){};
+	virtual void add_sample_bus(int p_at_pos = -1){};
+	virtual void move_sample_bus(int p_bus, int p_to_pos){};
+	virtual void set_sample_bus_volume_db(int p_bus, float p_volume_db){};
+	virtual void set_sample_bus_solo(int p_bus, bool p_enable){};
+	virtual void set_sample_bus_mute(int p_bus, bool p_enable){};
```
```suggestion
	}
	virtual void register_sample(const Ref<AudioSample> &p_sample) {}
	virtual void unregister_sample(const Ref<AudioSample> &p_sample) {}
	virtual void start_sample_playback(const Ref<AudioSamplePlayback> &p_playback) {}
	virtual void stop_sample_playback(const Ref<AudioSamplePlayback> &p_playback) {}
	virtual void set_sample_playback_pause(const Ref<AudioSamplePlayback> &p_playback, bool p_paused) {}
	virtual bool is_sample_playback_active(const Ref<AudioSamplePlayback> &p_playback) { return false; }
	virtual void update_sample_playback(const Ref<AudioSamplePlayback> &p_playback, const StringName &p_bus, float p_pan = 0.0f, float p_volume_db = 0.0f, float p_pitch_scale = 0.0f) {}

	virtual void set_sample_bus_count(int p_count) {}
	virtual void remove_sample_bus(int p_bus) {}
	virtual void add_sample_bus(int p_at_pos = -1) {}
	virtual void move_sample_bus(int p_bus, int p_to_pos) {}
	virtual void set_sample_bus_volume_db(int p_bus, float p_volume_db) {}
	virtual void set_sample_bus_solo(int p_bus, bool p_enable) {}
	virtual void set_sample_bus_mute(int p_bus, bool p_enable) {}
```

adamscott replied:
It's `clang-format` that made me change this.

AThousandShips replied:
Yes because you added `;` at the end, if you remove them this is correct

akien-mga commented:
Testing the current version on https://github.com/johncoffee/ngj-2024, it works really well!

On thing I noticed is that volume on the AudioStreamPlayer's doesn't seem to work. That's why if you test that demo, the footstep and attack sounds are much louder on web than on desktop. Is that a limitation of samples, or something that can be fixed? I know we can't do fancy bus effects but volume is quite important for most games IMO.

Calinou commented:
Is the configurable stereo separation strength supported as well? We have properties for this on a per-AudioStreamPlayer2D/3D node, as well as manual panning in AudioStreamPlayer and a project setting that acts as a global multiplier for stereo separation.

adamscott commented:
> Testing the current version on https://github.com/johncoffee/ngj-2024, it works really well!
> 
> On thing I noticed is that volume on the AudioStreamPlayer's doesn't seem to work. That's why if you test that demo, the footstep and attack sounds are much louder on web than on desktop. Is that a limitation of samples, or something that can be fixed? I know we can't do fancy bus effects but volume is quite important for most games IMO.

I'm on the volume issues!

adamscott commented:
> as well as manual panning in AudioStreamPlayer
How? I was never able to do this.

adamscott commented:
@akien-mga I think I fixed the volume issue.

Calinou commented:
> @Calinou How? I was never able to do this.

Nevermind, I thought we had this feature in AudioStreamPlayer, but we don't.

Calinou reviewed (commented):
Tested locally, it works as expected on a Samsung Galaxy Z Fold 4 on Firefox 125 and on a Samsung Galaxy Tab S9 Ultra on [Cromite](https://github.com/uazo/cromite) 124.

[Bromite](https://www.bromite.org/) 108 doesn't work (I get a `Could not load sample ...` error at load that prevents running the game altogether), but it's quite old by now. The main thread variant works correctly, but only when playing in the background: sound playback is heavily distorted when the page is focused.

Either way, it's not very important to have this working on an outdated browser, especially since Bromite was superseded by Cromite.

Remember to add descriptions for newly exposed classes/properties/methods in the class reference. Also, are there are any advantages to using sample playback on non-web platforms? If so, this should be documented.

adamscott commented:
> Remember to add descriptions for newly exposed classes/properties/methods in the class reference.

Yes! I'll add them for sure. Thanks for the reminder.

> Also, are there are any advantages to using sample playback on non-web platforms? If so, this should be documented.

Absolutely not. Sample playback for non-supported platforms does nothing, so no sound.

KoBeWi commented on scene/2d/audio_stream_player_2d.cpp:
```diff
@@ -219,6 +236,17 @@ void AudioStreamPlayer2D::play(float p_from_pos) {
 	setplayback = stream_playback;
 	setplay.set(p_from_pos);
+
+	// Sample handling.
+	if (stream_playback->get_is_sample()) {
+		Ref<AudioSamplePlayback> sample_playback = stream_playback->get_sample_playback();
+		sample_playback->offset = p_from_pos;
+		sample_playback->volume_db = get_volume_db();
+		sample_playback->position_mode = AudioSamplePlayback::AUDIO_SAMPLE_PLAYBACK_POSITION_2D;
+		sample_playback->bus = _get_actual_bus();
+
+		AudioServer::get_singleton()->start_sample_playback(sample_playback);
+	}
```
This can be moved to a method in AudioStreamPlayerInternal, with differences between players passed as parameters.

adamscott replied:
I think I can even get more low-level than that.

akien-mga commented:
> > Also, are there are any advantages to using sample playback on non-web platforms? If so, this should be documented.
> 
> Absolutely not. Sample playback for non-supported platforms does nothing, so no sound.

I'm a bit concerned about this when it comes to the UX of having a new property on audio players. We now have a property on each player, and a global setting, to confirmed whether to use streams or samples. But on any platform other than web choosing anything else than stream would break the project.

Wouldn't it be better to make this an implementation detail that's handled behind the scenes in the web platform, and not expose this at all to the user? Or do we need users to still be able to choose running specific AudioStreamPlayers as streams even on web?

adamscott commented:
> Or do we need users to still be able to choose running specific AudioStreamPlayers as streams even on web?

For dynamic streams, absolutely, as they are not supported currently as samples. Maybe we could add a ⚠️ emoji if we see that an incompatible setting has been selected?

akien-mga commented:
Tested the current PR with https://github.com/gdquest-demos/godot-4-3d-third-person-controller, it seems to break the music playback.

adamscott commented:
> Tested the current PR with https://github.com/gdquest-demos/godot-4-3d-third-person-controller, it seems to break the music playback.

Added the bug in my list of TODOs before merging: the autoplay property doesn't seem to trigger a play.

hakro commented:
I've tested all the games mentioned in the OP.
The sound worked pretty well for all of them on both FF 126 and Brave 1.66 on Linux.
Only Truck Town had a bit of crackling on both, but I think it might just be something with the game itself.
And yeah, as mentioned above, the autoplay does not trigger. I had to click once inside the game to start the sound.
Everything was pretty smooth on mobile as well.

adamscott commented:
> And yeah, as mentioned above, the autoplay does not trigger. I had to click once inside the game to start the sound.

That's a Web feature (called [autoplay blocking](https://developer.mozilla.org/en-US/docs/Web/Media/Autoplay_guide#autoplay_and_autoplay_blocking))! Sound doesn't play before an interaction event from the user is triggered.

I was rather talking about AudioStreamPlayer's internal "autoplay" feature.

MBCX commented:
I can't compile this branch on my Godot build, it errors at the end with this wall of text:
```bash
em++: warning: -pthread + ALLOW_MEMORY_GROWTH may run non-wasm code slowly, see https://github.com/WebAssembly/design/issues/1271 [-Wpthreads-mem-growth]
file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:3567
  var err = new SyntaxError(message);
            ^

SyntaxError: Unexpected token (11023:19)
  	static _samples = new Map();
                   ^

    at Parser.pp$4.raise (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:3567:13)
    at Parser.pp$9.unexpected (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:766:8)
    at Parser.pp$9.expect (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:760:26)
    at Parser.pp$5.parseMethod (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:3321:8)
    at Parser.pp$8.parseClassMethod (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:1527:35)
    at Parser.pp$8.parseClassElement (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:1485:10)
    at Parser.pp$8.parseClass (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:1403:24)
    at Parser.pp$5.parseExprAtom (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:2909:17)
    at Parser.pp$5.parseExprSubscripts (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:2709:19)
    at Parser.pp$5.parseMaybeUnary (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:2675:17) {
  pos: 396985,
  loc: Position { line: 11023, column: 19 },
  raisedAt: 396986
}
em++: error: '/home/mbcx/sources/emsdk/node/16.20.0_64bit/bin/node /home/mbcx/sources/emsdk/upstream/emscripten/tools/acorn-optimizer.mjs /tmp/emscripten_temp_wvhct4ms/godot.web.template_debug.wasm32.js growableHeap --closureFriendly -o /tmp/emscripten_temp_wvhct4ms/godot.web.template_debug.wasm32.jso1.js' failed (returned 1)
scons: *** [bin/godot.web.template_debug.wasm32.js] Error 1
scons: building terminated because of errors.
```

adamscott commented:
@MBCX Which version of emscripten are you using? I used the latest one, maybe it has an impact.

Calinou commented:
>  Only Truck Town had a bit of crackling on both, but I think it might just be something with the game itself.

How much FPS do you get when running the project? Stuttering might happen at low FPS because the engine sound pitch only changes once per frame, so having low FPS (below 30 or so) will cause discrete pitch changes to be noticeable. A lot of racing games run into this issue as it's difficult to fix.

MBCX commented:
> @MBCX Which version of emscripten are you using? I used the latest one, maybe it has an impact.

@adamscott I am using emscripten 3.1.56, I'm going to update to 3.1.59 and see if that fixes it.

MBCX commented:
@adamscott Nope, even after updating to latest emscripten (3.1.59), it still fails with same errors:
```bash
em++: warning: -pthread + ALLOW_MEMORY_GROWTH may run non-wasm code slowly, see https://github.com/WebAssembly/design/issues/1271 [-Wpthreads-mem-growth]
file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:3567
  var err = new SyntaxError(message);
            ^

SyntaxError: Unexpected token (11023:19)
  	static _samples = new Map();
                   ^

    at Parser.pp$4.raise (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:3567:13)
    at Parser.pp$9.unexpected (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:766:8)
    at Parser.pp$9.expect (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:760:26)
    at Parser.pp$5.parseMethod (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:3321:8)
    at Parser.pp$8.parseClassMethod (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:1527:35)
    at Parser.pp$8.parseClassElement (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:1485:10)
    at Parser.pp$8.parseClass (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:1403:24)
    at Parser.pp$5.parseExprAtom (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:2909:17)
    at Parser.pp$5.parseExprSubscripts (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:2709:19)
    at Parser.pp$5.parseMaybeUnary (file:///home/mbcx/sources/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.mjs:2675:17) {
  pos: 396985,
  loc: Position { line: 11023, column: 19 },
  raisedAt: 396986
}
em++: error: '/home/mbcx/sources/emsdk/node/16.20.0_64bit/bin/node /home/mbcx/sources/emsdk/upstream/emscripten/tools/acorn-optimizer.mjs /tmp/emscripten_temp_wvhct4ms/godot.web.template_debug.wasm32.js growableHeap --closureFriendly -o /tmp/emscripten_temp_wvhct4ms/godot.web.template_debug.wasm32.jso1.js' failed (returned 1)
scons: *** [bin/godot.web.template_debug.wasm32.js] Error 1
scons: building terminated because of errors.
```
```bash
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.59 (0e4c5994eb5b8defd38367a416d0703fd506ad81)
Copyright (C) 2014 the Emscripten authors (see AUTHORS.txt)
This is free and open source software under the MIT license.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

akien-mga commented:
@MBCX Could you share the exact build command you're using?

MBCX commented:
> @MBCX Could you share the exact build command you're using?

@akien-mga I build Godot in a automated manner through a script, this is the command that gets formed before executing:
```bash
SCRIPT_AES256_ENCRYPTION_KEY="<key>"

# Debug command (fails)
scons -j4 \ # ${THREADS_PER_JOB}
    platform="web"  precision="single" module_text_server_fb_enabled=yes \ # $common_options
    threads="yes" \ # $thread_mode
    target="template_debug"

# Release command (not tried yet)
scons -j4 \ # ${THREADS_PER_JOB}
    platform="web"  precision="single" module_text_server_fb_enabled=yes \ # $common_options
    threads="yes" \ # $thread_mode
    target="template_release" production="yes" use_closure_compiler="yes"
```

adamscott commented:
@MBCX It seems related to using the closure compiler. I'll take a look why it has issues with my new JavaScript code.

MBCX commented:
> @MBCX It seems related to using the closure compiler. I'll take a look why it has issues with my new JavaScript code.

Well, I haven't tried compiling with closure compiler, compiling template_debug is enough to trigger the bug

akien-mga commented:
I can reproduce the issue too with `scons p=web threads=yes target=template_debug`, I guess you just haven't compiled this specific branch with threads enabled @adamscott (since the whole point is to improve things without threads).

Fixing the style issues from CI would likely also have surfaced this in the CI build :)

adamscott commented:
Fixed the issue with closure-compiler. I made the "classes" prototypes instead.

akien-mga commented on scene/animation/animation_mixer.cpp:
```diff
@@ -33,7 +33,10 @@
 
 #include "core/config/engine.h"
 #include "core/config/project_settings.h"
+#include "scene/2d/audio_stream_player_2d.h"
+#include "scene/3d/audio_stream_player_3d.h"
 #include "scene/animation/animation_player.h"
+#include "scene/audio/audio_stream_player.h"
```
Those seem unused, and cause a compilation error:
https://github.com/godotengine/godot/actions/runs/9128230788/job/25100184935?pr=91382

Remember the `disable_3d=yes` SCons option which disables compilation of most 3D nodes. So including 3D nodes in non-3D ones needs using `#ifndef _3D_DISABLED` guards (but here it doesn't seem used in the end, which is better to ensure encapsulation).

AThousandShips replied:
This is still relevant

Maran23 commented:
When testing, I always get an exception in the console.

`Bus.count;` -> `return GodotAudio.Bus._buses.length;` -> `GodotAudio.Bus` is null?

```
tmp_js_export.js:10089 Uncaught (in promise) TypeError: Cannot read properties of null (reading '_buses')
    at Function.get (tmp_js_export.js:10089:29)
    at Object.GodotAudioClassesInit [as classesInit] (tmp_js_export.js:10116:8)
    at tmp_js_export.js:14275:16
    at tmp_js_export.js:16297:8
```

I don't do anything crazy with the audio buses:
![image](https://github.com/godotengine/godot/assets/66004280/18f793a0-3e20-474f-b534-5152990b2a38)

thygrrr commented:
The "Demo without sound issues" in the Blog Post that lead me here has no sound at all.
Chrome, Windows 11 Home, Version 125.0.6422.60 (Official Build) (64-bit)

AThousandShips commented on platform/web/audio_driver_web.cpp:
```diff
@@ -186,6 +187,194 @@ Error AudioDriverWeb::input_stop() {
+	constexpr int real_max_channels = AudioServer::MAX_CHANNELS_PER_BUS * 2;
+	PackedFloat32Array volume;
+	volume.resize(real_max_channels);
+	for (int i = 0; i < real_max_channels; i += 2) {
+		if (p_playback->volume_vector.size() == 0) {
+			volume.set(i, 0);
+			volume.set(i + 1, 0);
+		} else {
+			const AudioFrame &frame = p_playback->volume_vector.get(i / 2);
+			volume.set(i, frame.left);
+			volume.set(i + 1, frame.right);
+		}
+	}
+
+	godot_audio_sample_start(
+			itos(p_playback->get_instance_id()).utf8().get_data(),
+			itos(p_playback->stream->get_instance_id()).utf8().get_data(),
+			AudioServer::get_singleton()->get_bus_index(p_playback->bus),
+			p_playback->offset,
+			volume.ptrw(),
```
```suggestion
	float *volume_ptrw = volume.ptrw();
	for (int i = 0; i < real_max_channels; i += 2) {
		if (p_playback->volume_vector.is_empty()) {
			volume_ptrw[i] = 0;
			volume_ptrw[i + 1] = 0;
		} else {
			const AudioFrame &frame = p_playback->volume_vector[i / 2];
			volume_ptrw[i] = frame.left;
			volume_ptrw[i + 1] = frame.right;
		}
	}

	godot_audio_sample_start(
			itos(p_playback->get_instance_id()).utf8().get_data(),
			itos(p_playback->stream->get_instance_id()).utf8().get_data(),
			AudioServer::get_singleton()->get_bus_index(p_playback->bus),
			p_playback->offset,
			volume_ptrw,
```

AThousandShips commented on platform/web/audio_driver_web.cpp:
```diff
@@ -186,6 +187,194 @@ Error AudioDriverWeb::input_stop() {
+
+	PackedInt32Array buses;
+	buses.resize(p_bus_volumes.size());
+	PackedFloat32Array values;
+	values.resize(p_bus_volumes.size() * AudioServer::MAX_CHANNELS_PER_BUS * 2);
+
+	int idx = 0;
+	for (KeyValue<StringName, Vector<AudioFrame>> pair : p_bus_volumes) {
+		int bus_index = AudioServer::get_singleton()->get_bus_index(pair.key);
+		buses.set(idx, bus_index);
+		ERR_FAIL_COND(pair.value.size() != AudioServer::MAX_CHANNELS_PER_BUS);
+		for (int i = 0; i < real_max_channels; i += 2) {
+			const AudioFrame &frame = pair.value.get(i / 2);
+			values.set((idx * real_max_channels) + i, frame.left);
+			values.set((idx * real_max_channels) + i + 1, frame.right);
+		}
+		idx++;
+	}
+
+	godot_audio_sample_set_volumes_linear(
+			itos(p_playback->get_instance_id()).utf8().get_data(),
+			buses.ptrw(),
+			buses.size(),
+			values.ptrw(),
```
```suggestion
	buses_ptrw = buses.ptrw();
	PackedFloat32Array values;
	values.resize(p_bus_volumes.size() * AudioServer::MAX_CHANNELS_PER_BUS * 2);
	values_ptrw = values.ptrw();

	int idx = 0;
	for (KeyValue<StringName, Vector<AudioFrame>> pair : p_bus_volumes) {
		int bus_index = AudioServer::get_singleton()->get_bus_index(pair.key);
		buses_ptrw[idx] = bus_index;
		ERR_FAIL_COND(pair.value.size() != AudioServer::MAX_CHANNELS_PER_BUS);
		for (int i = 0; i < real_max_channels; i += 2) {
			const AudioFrame &frame = pair.value[i / 2];
			values_ptrw[(idx * real_max_channels) + i] = frame.left;
			values_ptrw[(idx * real_max_channels) + i + 1] = frame.right;
		}
		idx++;
	}

	godot_audio_sample_set_volumes_linear(
			itos(p_playback->get_instance_id()).utf8().get_data(),
			buses_ptrw,
			buses.size(),
			values_ptrw,
```

adamscott commented:
> When testing, I always get an exception in the console.
> 
> `Bus.count;` -> `return GodotAudio.Bus._buses.length;` -> `GodotAudio.Bus` is null?
> 
> ```
> tmp_js_export.js:10089 Uncaught (in promise) TypeError: Cannot read properties of null (reading '_buses')
>     at Function.get (tmp_js_export.js:10089:29)
>     at Object.GodotAudioClassesInit [as classesInit] (tmp_js_export.js:10116:8)
>     at tmp_js_export.js:14275:16
>     at tmp_js_export.js:16297:8
> ```
> 
> I don't do anything crazy with the audio buses: ![image](https://private-user-images.githubusercontent.com/66004280/331831665-18f793a0-3e20-474f-b534-5152990b2a38.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcwODU1NDYsIm5iZiI6MTcxNzA4NTI0NiwicGF0aCI6Ii82NjAwNDI4MC8zMzE4MzE2NjUtMThmNzkzYTAtM2UyMC00NzRmLWI1MzQtNTE1Mjk5MGIyYTM4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTMwVDE2MDcyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU1ZjIwNGE2MWE3NTA4NGI2MGM5ZDkzYTlmMTMwNzg3NzM4ZDk3MjZjOTM1YzcyNjhjMDVlNjYyMTJlODI3ZGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._kYV9ahp0RpGCESQPP-cVLV2UgvGHYOE8wmvbtWVmDc)

@Maran23 Can you reproduce the bug with my latest changes? If so, can you join a minimal reproducible project?

adamscott commented on platform/web/detect.py:
```diff
@@ -170,7 +170,7 @@ def configure(env: "SConsEnvironment"):
 
     # Closure compiler extern and support for ecmascript specs (const, let, etc).
-    env["ENV"]["EMCC_CLOSURE_ARGS"] = "--language_in ECMASCRIPT_2021"
+    env["ENV"]["EMCC_CLOSURE_ARGS"] = "--language_in UNSTABLE"
```
@Faless I changed the "language in" to `UNSTABLE` in order to support static class methods. Otherwise, it doesn't compile.

```
/var/folders/_5/6_c63ny97m554cs3yqq3djxh0000gn/T/emscripten_temp_u8l09o95/godot.web.template_debug.dev.wasm32.nothreads.js:9645:4: ERROR - [JSC_LANGUAGE_FEATURE] This language feature is only supported for UNSTABLE mode or better: Public class fields.
  9645|                 static _samples = new Map();
                        ^

/var/folders/_5/6_c63ny97m554cs3yqq3djxh0000gn/T/emscripten_temp_u8l09o95/godot.web.template_debug.dev.wasm32.nothreads.js:9854:4: ERROR - [JSC_LANGUAGE_FEATURE] This language feature is only supported for UNSTABLE mode or better: Public class fields.
  9854|                 static _sampleNodes = new Map();
                        ^

/var/folders/_5/6_c63ny97m554cs3yqq3djxh0000gn/T/emscripten_temp_u8l09o95/godot.web.template_debug.dev.wasm32.nothreads.js:10064:4: ERROR - [JSC_LANGUAGE_FEATURE] This language feature is only supported for UNSTABLE mode or better: Public class fields.
  10064|                static _buses = [];
                        ^

/var/folders/_5/6_c63ny97m554cs3yqq3djxh0000gn/T/emscripten_temp_u8l09o95/godot.web.template_debug.dev.wasm32.nothreads.js:10066:4: ERROR - [JSC_LANGUAGE_FEATURE] This language feature is only supported for UNSTABLE mode or better: Public class fields.
  10066|                static _busSolo = null;
                        ^
```

adamscott replied:
> `UNSTABLE` doesn't seem something we should target in stable release though.
> 
> Do we really need all those? None of the other libraries needs to use the `static` keyword. Why not use the usual:
> 
> ```
> const Lib = {
>   $Lib: {
>     myProp: 1
>   },
>   godot_js_get_my_prop() {
>     return Lib.myProp;
>   },
> };
> 
> mergeInto(LibraryManager.library, Lib);
> ```

You're right! I reverted back the target.

Faless commented on platform/web/js/libs/library_godot_audio.d.ts:
```diff
@@ -0,0 +1,158 @@
+export type PositionMode = "none" | "2D" | "3D";
```
Was this file committed on purpose? I don't understand why we need type definitions for interoperability with TypeScript for an internal library.

adamscott replied:
Yes. It adds typing for easier coding. I'll review the code to make sure that it's still relevant.

Faless commented:
Is it enforced by CI somehow? E.g. is the type definition used by the
Closure Compiler (docs doesn't seem to mention typescript imports support).

And if not, how are going to make sure the type definition actually matches
the JS code?

On Sun, Jun 2, 2024, 07:34 Adam Scott ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In platform/web/js/libs/library_godot_audio.d.ts
> <https://github.com/godotengine/godot/pull/91382#discussion_r1623330140>:
>
> > @@ -0,0 +1,158 @@
> +export type PositionMode = "none" | "2D" | "3D";
>
> Yes. It adds typing for easier coding. I'll review the code to make sure
> that it's still relevant.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/91382#discussion_r1623330140>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAM4C3XZM5DKLNB3ECJ2IVTZFKVFLAVCNFSM6AAAAABHBFL5QWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDAOJSGIYTQNZYGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Faless commented:
As far as I can tell (see https://github.com/jsdoc/jsdoc/issues/1929 and
https://github.com/jsdoc/jsdoc/issues/1645 ) it seems that the jsdoc syntax
used to import from TS type definition is a VSCode dialect.
I'm not sure the CC supports it, but we should be able to provide types
using standard JSDoc (or a dialect supported by the CC).
See
https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler

On Sun, Jun 2, 2024, 09:06 Fabio Alessandrelli <
***@***.***> wrote:

> Is it enforced by CI somehow? E.g. is the type definition used by the
> Closure Compiler (docs doesn't seem to mention typescript imports support).
>
> And if not, how are going to make sure the type definition actually
> matches the JS code?
>
> On Sun, Jun 2, 2024, 07:34 Adam Scott ***@***.***> wrote:
>
>> ***@***.**** commented on this pull request.
>> ------------------------------
>>
>> In platform/web/js/libs/library_godot_audio.d.ts
>> <https://github.com/godotengine/godot/pull/91382#discussion_r1623330140>:
>>
>> > @@ -0,0 +1,158 @@
>> +export type PositionMode = "none" | "2D" | "3D";
>>
>> Yes. It adds typing for easier coding. I'll review the code to make sure
>> that it's still relevant.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/godotengine/godot/pull/91382#discussion_r1623330140>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAM4C3XZM5DKLNB3ECJ2IVTZFKVFLAVCNFSM6AAAAABHBFL5QWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDAOJSGIYTQNZYGY>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

adamscott commented:
> Is it enforced by CI somehow? E.g. is the type definition used by the Closure Compiler (docs doesn't seem to mention typescript imports support). And if not, how are going to make sure the type definition actually matches the JS code?

I removed the `.d.ts` file and replaced everything with Closure Compiler compatible JSDoc.

adamscott commented:
For some reason, when try to open up the advanced importer for audio files, it triggers an infinite loop. Not present on master, but I don't know what changed to create this bug.

patwork commented:
I have created new scene with two AudioStreamPlayer2D nodes activated by mouse over signal from Area2D.

![scene](https://github.com/godotengine/godot/assets/224215/5dec62f0-03a0-40a3-b46a-babb90bdaae2)

```
func _on_mouse_entered() -> void:
	player.play()
func _on_mouse_exited() -> void:
	player.stop()
```

When exported to web, the sounds are played when the mouse hovers over the icons. Everything works when I cancel playback by moving the mouse away. If I let the sample play to the end, an error appears:

```
Uncaught ReferenceError: Could not find sample node "-9223371954411535061"
    getSampleNode http://localhost:8060/tmp_js_export.js:9846
    sample_set_volumes_linear http://localhost:8060/tmp_js_export.js:10622
    _godot_audio_sample_set_volumes_linear http://localhost:8060/tmp_js_export.js:10817
    callUserCallback http://localhost:8060/tmp_js_export.js:6634
    runIter http://localhost:8060/tmp_js_export.js:6709
    Browser_mainLoop_runner http://localhost:8060/tmp_js_export.js:6595
    requestAnimationFrame http://localhost:8060/tmp_js_export.js:7004
    Browser_mainLoop_scheduler_rAF http://localhost:8060/tmp_js_export.js:5888
    Browser_mainLoop_runner http://localhost:8060/tmp_js_export.js:6608
    requestAnimationFrame http://localhost:8060/tmp_js_export.js:7004
    Browser_mainLoop_scheduler_rAF http://localhost:8060/tmp_js_export.js:5888
    Browser_mainLoop_runner http://localhost:8060/tmp_js_export.js:6608
    requestAnimationFrame http://localhost:8060/tmp_js_export.js:7004
    Browser_mainLoop_scheduler_rAF http://localhost:8060/tmp_js_export.js:5888
    Browser_mainLoop_runner http://localhost:8060/tmp_js_export.js:6608
    requestAnimationFrame http://localhost:8060/tmp_js_export.js:7004
    Browser_mainLoop_scheduler_rAF http://localhost:8060/tmp_js_export.js:5888
    Browser_mainLoop_runner http://localhost:8060/tmp_js_export.js:6608
    requestAnimationFrame http://localhost:8060/tmp_js_export.js:7004
    Browser_mainLoop_scheduler_rAF http://localhost:8060/tmp_js_export.js:5888
    Browser_mainLoop_runner http://localhost:8060/tmp_js_export.js:6608
    requestAnimationFrame http://localhost:8060/tmp_js_export.js:7004
    Browser_mainLoop_scheduler_rAF http://localhost:8060/tmp_js_export.js:5888
    Browser_mainLoop_runner http://localhost:8060/tmp_js_export.js:6608
    requestAnimationFrame http://localhost:8060/tmp_js_export.js:7004
    Browser_mainLoop_scheduler_rAF http://localhost:8060/tmp_js_export.js:5888
    Browser_mainLoop_runner http://localhost:8060/tmp_js_export.js:6608
    requestAnimationFrame http://localhost:8060/tmp_js_export.js:7004
    Browser_mainLoop_scheduler_rAF http://localhost:8060/tmp_js_export.js:5888
    Browser_mainLoop_runner http://localhost:8060/tmp_js_export.js:6608
    requestAnimationFrame http://localhost:8060/tmp_js_export.js:7004
    Browser_mainLoop_scheduler_rAF http://localhost:8060/tmp_js_export.js:5888
    Browser_mainLoop_runner http://localhost:8060/tmp_js_export.js:6608
    requestAnimationFrame http://localhost:8060/tmp_js_export.js:7004
    Browser_mainLoop_scheduler_rAF http://localhost:8060/tmp_js_export.js:5888
    Browser_mainLoop_runner http://localhost:8060/tmp_js_export.js:6608
    requestAnimationFrame http://localhost:8060/tmp_js_export.js:7004
    Browser_mainLoop_scheduler_rAF http://localhost:8060/tmp_js_export.js:5888
    Browser_mainLoop_runner http://localhost:8060/tmp_js_export.js:6608
    requestAnimationFrame http://localhost:8060/tmp_js_export.js:7004
    Browser_mainLoop_scheduler_rAF http://localhost:8060/tmp_js_export.js:5888
    Browser_mainLoop_runner http://localhost:8060/tmp_js_export.js:6608
    requestAnimationFrame http://localhost:8060/tmp_js_export.js:7004
    Browser_mainLoop_scheduler_rAF http://localhost:8060/tmp_js_export.js:5888
    Browser_mainLoop_runner http://localhost:8060/tmp_js_export.js:6608
tmp_js_export.js:9846:12
```

❗ ALSO: despite using AudioStreamPlayer2D and setting `panning_strength` to “3”, I can't hear the left-right sound positioning effect (it is very clear in Windows export)

AThousandShips commented on scene/resources/audio_stream_polyphonic.compat.inc:
```diff
@@ -0,0 +1,41 @@
+#ifndef DISABLE_DEPRECATED
+
+AudioStreamPlaybackPolyphonic::ID AudioStreamPlaybackPolyphonic::_play_stream_bind_compat_91382(const Ref<AudioStream> &p_stream, float p_from_offset, float p_volume_db, float p_pitch_scale) {
+	return play_stream(p_stream, p_from_offset, p_volume_db, p_pitch_scale);
```
```suggestion
	return play_stream(p_stream, p_from_offset, p_volume_db, p_pitch_scale, AudioServer::PlaybackType::PLAYBACK_TYPE_DEFAULT, SceneStringName(Master));
```
Arguments should be explicit in a compatibility bind (it also needs to be added to the expected file

adamscott commented:
@dsnopek Do I need to add something in godot-cpp?
https://github.com/godotengine/godot/actions/runs/9437714281/job/25993895574?pr=91382

AThousandShips commented:
You need to update the `misc/extension_api_validation/4.2-stable.expected` file, like so:
```

GH-91382
--------
Validate extension JSON: Error: Field 'classes/AudioStreamPlaybackPolyphonic/methods/play_stream/arguments': size changed value in new API, from 4 to 6.

SOME NOTE, compatibility method registered.
```
At the end of that file

Faless commented on platform/web/audio_driver_web.cpp:
```diff
@@ -105,7 +106,10 @@ void AudioDriverWeb::_audio_driver_capture(int p_from, int p_samples) {
 	int latency = Engine::get_singleton()->get_audio_output_latency();
 	if (!audio_context.inited) {
-		audio_context.mix_rate = _get_configured_mix_rate();
+		int mix_rate = _get_configured_mix_rate();
+		audio_context.mix_rate = mix_rate == 0
+				? godot_audio_get_browser_mix_rate()
+				: mix_rate;
```
This change is not really needed because `audio_context.mix_rate` is passed as a pointer to `godot_audio_init` which sets it to the actual resulting `AudioContext.sampleRate`.

```suggestion
		// If 0, it will be set to the browser preferred value after calling godot_audio_init.
		audio_context.mix_rate = _get_configured_mix_rate();
```

While it might be tempting to have a `godot_audio_get_browser_mix_rate()` function, this forces us to create a new AudioContext which:
- Cause an extra console warning if audiocontext was not allowed to start (not a result of a user action)
- Uses a part of the spec (opening multiple `AudioContext`s) which is [under-standardized](https://developer.mozilla.org/en-US/docs/Web/API/AudioContext/AudioContext#usage_notes) (especially given we can't await for the first one to close).

In practice, this would likely be "safe enough", but I've learned not to trust vendors in maintaining compatibility with poorly standardized behaviors.

So, given we don't seem to need that function anywhere else, I suggest to remove it completely.

adamscott replied:
@Faless I added this recently because it causes issues on Safari otherwise (the other browsers are fine).

adamscott replied:
I'll find a way to only use one `AudioContext`.

Faless replied:
> I added this recently because it causes issues on Safari otherwise (the other browsers are fine).

In theory this change should not affect browser behavior.

With the change in the PR:
- We create an audio context without specifying any parameter.
- We read the assigned mix rate.
- Create a the true audio context with this "default" mix rate.
- Return the resulting audio context mix rate (sampleRate) to C++ as an out parameter.

With my change (and as currently in upstream):
- We create an audio context specifying the mix rate *only if* it's not zero.
- Return the resulting audio context mix rate (sampleRate) to C++ as an out parameter.

If this change affects Safari, it might be an issue with the way we set/read the sample rate. Try printing the values on both C++ `audio_context.mix_rate` and JS side `ctx.sampleRate`.

If all of the above seems to be correctly set (and it's not a bug in some of my old code :sweat_smile: ), we might be hitting different internal code path in Safari (when the sampleRate is and is not specified). Browsers usually greatly prefer not being forced a specific mix rate, but Safari has always been the weird of the bunch, and I currently can't test that specifically.

Faless commented on platform/web/godot_audio.h:
```diff
@@ -44,9 +44,31 @@ extern int godot_audio_has_script_processor();
 extern int godot_audio_init(int *p_mix_rate, int p_latency, void (*_state_cb)(int), void (*_latency_cb)(float));
 extern void godot_audio_resume();
 
+extern int godot_audio_get_browser_mix_rate();
+
```
```suggestion
```

Faless commented on platform/web/js/libs/library_godot_audio.js:
```diff
@@ -190,6 +1375,13 @@ const GodotAudio = {
 		}
 	},
 
+	godot_audio_get_browser_mix_rate__proxy: 'sync',
+	godot_audio_get_browser_mix_rate__sig: 'i',
+	godot_audio_get_browser_mix_rate: function () {
+		const ctx = new AudioContext();
+		return ctx.sampleRate;
+	},
+
```
```suggestion
```

Faless reviewed (approved):
Amazing work :rocket:  :trophy: :cookie: !

I've tested this for a while and with all (I hope) possible combinations of threads/nothreads/dlink/nodlink/sample/stream (emcc 3.1.39).

It's working great on all of them in my tests.

I've added one last suggestion (and long explainer), but it's otherwise good to go.

Thank you so much :blue_heart:  for the incredible work and effort in getting this so far!

AThousandShips reviewed (approved):
Haven't tested but the code I've read looks good, still needs that api validation addition to pass CI though, the docs look good as well 🎉

Calinou commented:
When I run https://adamscott.github.io/2d-platformer-demo-main-thread-samples/ in Firefox 125, I get this error every time a coin is picked up, and I don't hear its coin pickup sound:

```
USER WARNING: /root/Game/Level/Coins/CoinsArc1/Coin2/Pickup is trying to play a sample from a stream that cannot be sampled. [index.js:16331:18](https://adamscott.github.io/2d-platformer-demo-main-thread-samples/index.js)
   at: play_basic (scene/audio/audio_stream_player_internal.cpp:158) [index.js:16331:18](https://adamscott.github.io/2d-platformer-demo-main-thread-samples/index.js)

class of stream: AudioStreamPolyphonic
```

Maran23 commented:
With the newest version, I got another error:
![image](https://github.com/godotengine/godot/assets/66004280/a87279d1-495f-49c3-b2e7-c00a64838cd9)

This happens after a while, it looks like when the game is paused (stacktrace also seems to indicate this). Will update this comment if I know more.

adamscott commented:
> With the newest version, I got another error: ![image](https://private-user-images.githubusercontent.com/66004280/339069096-a87279d1-495f-49c3-b2e7-c00a64838cd9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyODYzNzQsIm5iZiI6MTcxODI4NjA3NCwicGF0aCI6Ii82NjAwNDI4MC8zMzkwNjkwOTYtYTg3Mjc5ZDEtNDk1Zi00OWMzLWIyZTctYzAwYTY0ODM4Y2Q5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEzVDEzNDExNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU0YTkxOTlmOGIzZmY0YTJjMTVlNmFjNWU1YTA3MGJiODg1OWU1NGZhNjk3ZTkxYjk4YjVlY2QwYWQ3ZGZiMGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Hi-Rnwg42wd84_om61TuET9uVnSDAKk0qFgFhB2avYo)
> 
> This happens after a while, it looks like when the game is unpaused (stacktrace also seems to indicate this). Will update this comment if I know more.

@Maran23 The stack trace indicate that Godot wants to pause a sample, not that the game is paused.

adamscott commented:
> When exported to web, the sounds are played when the mouse hovers over the icons. Everything works when I cancel playback by moving the mouse away. If I let the sample play to the end, an error appears:
> 
> ```
> Uncaught ReferenceError: Could not find sample node "-9223371954411535061"
>     getSampleNode http://localhost:8060/tmp_js_export.js:9846
>     sample_set_volumes_linear http://localhost:8060/tmp_js_export.js:10622
>     _godot_audio_sample_set_volumes_linear http://localhost:8060/tmp_js_export.js:10817
> ```

@patwork Should be fixed!

adamscott commented:
@Maran23 Can you test again? I just pushed a fix that should resolve your issue.

AThousandShips commented:
Ouch it seems that the bindings won't work correctly generating string name arguments because of syntax, I don't think we can solve that in this so might be better to make the default empty and have it default to using `Master` if the name is empty, or similar

AThousandShips commented:
Got a fix that should work to help with that one, pushing to godot-cpp right now and we can see if we can't get it working

dsnopek commented:
PR https://github.com/godotengine/godot-cpp/pull/1487 was cherry-picked, so tests here _should_ pass. I just kicked off another test run, so we'll see!

akien-mga commented on platform/web/audio_driver_web.cpp:
```diff
@@ -186,6 +187,192 @@ Error AudioDriverWeb::input_stop() {
+	ERR_FAIL_COND_MSG(p_sample.is_null(), "Parameter p_sample is null.");
+	ERR_FAIL_COND_MSG(p_sample->stream.is_null(), "Parameter p_sample->stream is null.");
+
+	String loop_mode;
+	switch (p_sample->loop_mode) {
+		case AudioSample::LoopMode::LOOP_DISABLED: {
+			loop_mode = "disabled";
+		} break;
+
+		case AudioSample::LoopMode::LOOP_FORWARD: {
+			loop_mode = "forward";
+		} break;
+
+		case AudioSample::LoopMode::LOOP_PINGPONG: {
+			loop_mode = "pingpong";
+		} break;
+
+		case AudioSample::LoopMode::LOOP_BACKWARD: {
+			loop_mode = "backward";
+		} break;
+	}
```
Isn't it a bit wasteful to convert the enum to a string? Wouldn't it be simpler to pass an int and just replicate the enum on the .js side?

adamscott replied:
To maintain, it will be much easier (as there's no automatic map between the C++ enum and the JS one). And in JS, it's not rare to pass strings as enum values. So even if it's a bit wasteful, I suggest to keep it that way.

akien-mga commented on platform/web/audio_driver_web.cpp:
```diff
@@ -186,6 +187,192 @@ Error AudioDriverWeb::input_stop() {
+	}
+
+	double length = p_sample->stream->get_length();
+	switch (p_sample->loop_mode) {
+		case AudioSample::LoopMode::LOOP_FORWARD:
+		case AudioSample::LoopMode::LOOP_PINGPONG:
+		case AudioSample::LoopMode::LOOP_BACKWARD: {
+			// TODO: Calculate length based on `loop_begin` / `loop_end`.
+		} break;
```
I assume that TODO is still valid, if so I'd suggest opening an issue to keep track of it.

adamscott replied:
Finally, I think it was leftover code. I removed it. Thanks!

akien-mga commented on scene/resources/audio_stream_polyphonic.cpp:
```diff
@@ -260,8 +294,24 @@ void AudioStreamPlaybackPolyphonic::stop_stream(ID p_stream_id) {
+
 void AudioStreamPlaybackPolyphonic::_bind_methods() {
-	ClassDB::bind_method(D_METHOD("play_stream", "stream", "from_offset", "volume_db", "pitch_scale"), &AudioStreamPlaybackPolyphonic::play_stream, DEFVAL(0), DEFVAL(0), DEFVAL(1.0));
+	ClassDB::bind_method(D_METHOD("play_stream", "stream", "from_offset", "volume_db", "pitch_scale", "playback_type", "bus"), &AudioStreamPlaybackPolyphonic::play_stream, DEFVAL(0), DEFVAL(0), DEFVAL(1.0), DEFVAL(0), DEFVAL(SNAME("Master")));
```
```suggestion
	ClassDB::bind_method(D_METHOD("play_stream", "stream", "from_offset", "volume_db", "pitch_scale", "playback_type", "bus"), &AudioStreamPlaybackPolyphonic::play_stream, DEFVAL(0), DEFVAL(0), DEFVAL(1.0), DEFVAL(0), DEFVAL(SceneStringName(Master)));
```

akien-mga commented on doc/classes/AudioStreamPlayback.xml:
```diff
@@ -79,5 +79,18 @@
 				Overridable method. Called whenever the audio stream is mixed if the playback is active and [method AudioServer.set_enable_tagging_used_audio_streams] has been set to [code]true[/code]. Editor plugins may use this method to "tag" the current position along the audio stream and display it in a preview.
 			</description>
 		</method>
+		<method name="get_sample_playback" qualifiers="const" experimental="">
+			<return type="AudioSamplePlayback" />
+			<description>
+				Returns the meta class associated with this [AudioStreamPlayback] for playing back the audio sample of this stream.
+			</description>
+		</method>
+		<method name="set_sample_playback" experimental="">
+			<return type="void" />
+			<param index="0" name="playback_sample" type="AudioSamplePlayback" />
+			<description>
+				Associates the meta class to this [AudioStreamPlayback] for playing back the audio sample of this stream.
+			</description>
+		</method>
```
It's not clear to me what "meta class" means in this context. It's also an Object instance and not a class itself, no?

Overall the subtleties of streams vs samples are likely hard to convey through just API descriptions, especially as most users won't know which is which and why they should care, so we'll probably want to cover this in a dedicated docs page if there's a need for some users to handle this themselves.

akien-mga reviewed (commented):
Looks good to me overall, let's go!

Calinou commented:
https://adamscott.github.io/2d-platformer-demo-main-thread-samples/ now works great on both Firefox and Chromium, although I can still see the error message in the console every time I pick up a coin. The coin pickup sound still plays correctly though.

Calinou commented on doc/classes/AudioServer.xml:
```diff
@@ -344,5 +359,18 @@
+			Force the playback to be considered as a stream.
+		</constant>
+		<constant name="PLAYBACK_TYPE_SAMPLE" value="2" enum="PlaybackType" experimental="">
+			Force the playback to be considered as a sample.
+			Be careful, as not every platform supports samples.
```
```suggestion
			Force the playback to be considered as a sample.
			[b]Note:[/b] Only supported on the web platform. Other platforms will fall back to [constant PLAYBACK_TYPE_STREAM].
```

(Is the fallback behavior correct?)

adamscott replied:
I could maybe introduce a fallback behavior, but currently, the sound just doesn't play (it should display an error, though).

Mickeon replied:
Crackling* by the way.

Calinou commented on doc/classes/AudioStream.xml:
```diff
@@ -69,6 +81,12 @@
+		<method name="is_meta_stream" qualifiers="const">
+			<return type="bool" />
+			<description>
+				Returns if the stream is a collection of others streams.
```
```suggestion
				Returns if the stream is a collection of other streams.
```

Calinou commented on doc/classes/ProjectSettings.xml:
```diff
@@ -409,6 +409,14 @@
+			The default value is set to [code]Stream[/code], as most platforms have no issues mixing streams.
+		</member>
+		<member name="audio/general/default_playback_type.web" type="int" setter="" getter="" default="1" experimental="">
+			Specifies the default playback type of the Web platform.
+			The default value is set to [code]Sample[/code] as the Web platform (especially when exporting a single-threaded game) is not suited to mix audio streams outside of the Web Audio API.
```
```suggestion
			Specifies the default playback type of the Web platform.
			The default value is set to [code]Sample[/code] as the Web platform (especially when exporting a single-threaded game) is not suited to mix audio streams outside of the Web Audio API. Forcing [code]Stream[/code] on the Web platform will likely result in high audio latency and potential audio cracking, especially if running without threads enabled.
```

Mickeon replied:
Crackling*.
"will **likely** result in high **audio** latency and **potential** **audio** cracking" and similar repetitive wording adjustments (perhaps this is nitpicky).
```suggestion
			Specifies the default playback type of the Web platform.
			The default value is set to [b]Sample[/b] as the Web platform (especially when exporting a single-threaded game) is not suited to mix audio streams outside of the Web Audio API. [b]Sample[/b] allows for lower latency on the web platform at the cost of flexibility ([AudioEffect]s are not supported).
			[b]Warning:[/b] Forcing [b]Stream[/b] on the Web platform may cause high audio latency and crackling, especially when exporting a multi-threaded game.
```

Calinou commented on doc/classes/ProjectSettings.xml:
```diff
@@ -409,6 +409,14 @@
 		</member>
+		<member name="audio/general/default_playback_type" type="int" setter="" getter="" default="0" experimental="">
+			Specifies the default playback type of the platform.
+			The default value is set to [code]Stream[/code], as most platforms have no issues mixing streams.
```
```suggestion
			The default value is set to [b]Stream[/b], as most platforms have no issues mixing streams.
```

Mickeon commented on doc/classes/ProjectSettings.xml:
```diff
@@ -409,6 +409,15 @@
+			The default value is set to [b]Stream[/b], as most platforms have no issues mixing streams.
+		</member>
+		<member name="audio/general/default_playback_type.web" type="int" setter="" getter="" default="1" experimental="">
+			Specifies the default playback type of the Web platform.
+			The default value is set to [b]Sample[/b] as the Web platform (especially when exporting a single-threaded game) is not suited to mix audio streams outside of the Web Audio API. [b]Sample[/b] allows for lower latency on the web platform at the cost of flexibility ([AudioEffect]s are not supported).
+			[b]Warning:[/b] Forcing [b]Stream[/b] on the Web platform will likely result in high audio latency and potential audio cracking, especially if running without threads enabled.
```
See https://github.com/godotengine/godot/pull/91382#discussion_r1643331898 (I should've made it a separate review in hindsight)
```suggestion
Specifies the default playback type of the Web platform.
			The default value is set to [b]Sample[/b] as the Web platform is not suited to mix audio streams outside of the Web Audio API, especially when exporting a single-threaded game. [b]Sample[/b] allows for lower latency on the web platform at the cost of flexibility ([AudioEffect]s are not supported).
			[b]Warning:[/b] Forcing [b]Stream[/b] on the Web platform may cause high audio latency and crackling, especially when exporting a multi-threaded game.
```

Mickeon commented on doc/classes/AudioServer.xml:
```diff
@@ -198,6 +206,13 @@
+			<return type="void" />
+			<param index="0" name="stream" type="AudioStream" />
+			<description>
+				Forces the registration of a stream as a sample. Lag spikes may occur when calling this method, especially on single-threaded builds. It is suggested to call this method while loading assets, where the lag spike could be masked, instead of relying on registering the sample when it needs to be played.
```
These kind of long-winded sentences snuggly fit within a note because they:
- Do not describe the direct effect of the method
- Provide potential solutions or workarounds
```suggestion
				Forces the registration of a stream as a sample.
				[b]Note:[/b] Lag spikes may occur when calling this method, especially on single-threaded builds. It is suggested to call this method while loading assets, where the lag spike could be masked, instead of registering the sample right before it needs to be played.
```

Maran23 commented:
Sorry for the delay. Tested right now with then newest code, I got another error now when unpausing the game (and therefore all audio as well):
![image](https://github.com/godotengine/godot/assets/66004280/905e7d32-4034-4e8b-9dd9-fd5177ec8c9c)
![image](https://github.com/godotengine/godot/assets/66004280/16a0bec8-d6b3-41ec-aa6b-35ea5610f96f)

akien-mga commented:
Awesome work @adamscott!

Let's merge now to get wider testing, and fix the remaining issues in follow up PRs.

adamscott commented:
> Sorry for the delay. Tested right now with then newest code, I got another error now when unpausing the game (and therefore all audio as well): ![image](https://private-user-images.githubusercontent.com/66004280/340663733-905e7d32-4034-4e8b-9dd9-fd5177ec8c9c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg4MDIyMDUsIm5iZiI6MTcxODgwMTkwNSwicGF0aCI6Ii82NjAwNDI4MC8zNDA2NjM3MzMtOTA1ZTdkMzItNDAzNC00ZThiLTlkZDktZmQ1MTc3ZWM4YzljLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE5VDEyNTgyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI3NzdlMTY4OTBhY2JlZDVlYmNlY2ZjOTk3MmEzZjJkMGVjY2FjOGY2NDZmYzRlMGVlMTgwM2IyNTY0YTU2NTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Z4hY6dIp3wSEbWr92IMMoI2ZlVsW6_4e8Pt9tg8mSNc) ![image](https://private-user-images.githubusercontent.com/66004280/340663858-16a0bec8-d6b3-41ec-aa6b-35ea5610f96f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg4MDIyMDUsIm5iZiI6MTcxODgwMTkwNSwicGF0aCI6Ii82NjAwNDI4MC8zNDA2NjM4NTgtMTZhMGJlYzgtZDZiMy00MWVjLWFhNmItMzVlYTU2MTBmOTZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE5VDEyNTgyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk1MjA1MjFhMDcxYWFlZWU3MDY3MWMxMjg4YzJhYTIwYjJmZGRjZDZmYjVmNzBmMGJkZWIzYmE5ZmU5ZDRjYTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.aRSy9tpJKthwwDqfPj2To8Qs1o5LI-j5pIQF0lN8s_I)

I'll create an issue in the tracker. I'll try to reproduce this on my side ASAP.

JunShiozawa commented:
@adamscott 
I will try to backport this PR to Godot 3.x.

I need WebGL 1.0, so
